### PR TITLE
Mypy type errors fix

### PR DESCRIPTION
**Pull request by Domejko**
## Description

Fixed all type errors in *settings/* folder reducing there number from 384 to 369 errors.

Edit:
Part of files in *backend/* also have been fixed and errors count is reduced to 354.

Edit:
Mypy error count reduced from 354 to 313.

Created *utils/* folder, refactored **utils.py** to different files and updated imports accordingly. 
Added new subclass of HttpRequest reduce type errors as suggested in [django-stubs](https://github.com/typeddjango/django-stubs) documentation.

Edit:
- Resolved merge conflict due to #339 update
- Moved `event_bridge_scheduler` inside the function to call when needed due to 
```
ValueError: If using schedules, the variables MUST be set.
```
- as in #321 

# Checklist

- [x] Ran the [Black Formatter](https://docs.strelix.org/MyFinances/#/how-to-contribute?id=test-and-lint) and
  [djLint-er](https://docs.strelix.org/MyFinances/#/how-to-contribute?id=test-and-lint) on any new code
  (checks
  will
  fail without)
- [x] Made any changes or additions to the documentation _where required_
- [x] Changes generate no new warnings/errors
- [x] New and existing [unit tests](https://docs.strelix.org/MyFinances/#/how-to-contribute?id=test-and-lint) pass locally with my
  changes

## What type of PR is this?

- 🐛 Bug Fix
- ♻️ Code Refactor

## Added/updated tests?

- 🙅 no, because they aren't needed

## Related PRs, Issues etc
- Related Issue #319

### CODE REVIEW

File: backend/types/emails.py
Comment (TreyWW) [inline backend/types/emails.py:17]: ConfigurationSetName is optional though, so this breaks the actual functionality (type hints anyways)

File: backend/views/core/invoices/edit.py
Comment (TreyWW) [inline backend/views/core/invoices/edit.py:143]: This is different functionality though, the htmx module covers more than just Hx-Request

File: pyproject.toml
Comment (TreyWW) [inline pyproject.toml:139]: Why are we adding these to an override though, shouldn't we just slowly fix the errors rather than just ignore them? For django_components that makes sense but internal code we should fix rather than hide (in my opinion)

File: settings/helpers.py
Comment (TreyWW) [inline settings/helpers.py:108]: This should always be a string though

File: settings/local_settings.py
Comment (TreyWW) [inline settings/local_settings.py:20]: This should be a dict not "Any"

File: settings/local_settings.py
Comment (TreyWW) [inline settings/local_settings.py:51]: Ideally this should just be list[str] and never allow None

File: settings/prod_settings.py
Comment (TreyWW) [inline settings/prod_settings.py:43]: Ideally this should just be list[str] and never allow None

**Review by TreyWW — CHANGES_REQUESTED** [review]
There's a few minor issues I found with this PR but overall well done! Thanks for the contribution, would appreciate some changes made though :)

File: backend/types/emails.py
Comment (Domejko) [reply backend/types/emails.py:17]: I did changed that because `SESV2Client` expects `ConfigurationSetName` to be only of a type `str` and `mypy` complains about it. So as it need to be kept as `str | None = None` we can set `# type: ignore[arg-type]` or use in *settings/helpers.py* on line 132:
```
ConfigurationSetName=data.ConfigurationSetName or "",
```
insted of just
```
ConfigurationSetName=data.ConfigurationSetName,
```

Comment (TreyWW) [reply backend/types/emails.py:17]: Yeah that works

File: backend/views/core/invoices/edit.py
Comment (Domejko) [reply backend/views/core/invoices/edit.py:143]: You are sure that this line is working properly ? Because as far as I have been checking in Django HttpRequest source code and docs there is no `htmx` method and `mypy` along with `python` are complaining abut lack lack of any reference. And module from *backend/types/htmx.py* have not been used. 

I think we should use then in `edit_invoice` function `HtmxHttpRequest` from *backend/types/htmx.py* instead of regular *HttpRequest* because then it's all good.

Comment (TreyWW) [reply backend/views/core/invoices/edit.py:143]: We're using [Django-htmx](https://pypi.org/project/django-htmx/ ) is a library that includes .htmx in middleware

File: pyproject.toml
Comment (Domejko) [reply pyproject.toml:139]: I'm adding them because this part I would like to leave on the end. Creating stubs for all this modules will take some time and I thought that it will be better to first have other errors cleared since they are less troublesome and numbers will go down quicker to acceptable level.

Comment (TreyWW) [reply pyproject.toml:139]: Okay cool

File: settings/helpers.py
Comment (Domejko) [reply settings/helpers.py:108]: True but he sees that in `SingleTemplatedEmailContent` `template_data` is set to `dict | str` and is trowing an error suggesting a type of `dict[Any, Any] | str | None` and `reveal_type` is suggesting the same. So if we can't change or extend type of any of them solution that I see at the moment is to use `# type: ignore[assignment]` OR leave it without type annotation since it's named `data_str` and is used only once.

File: settings/prod_settings.py
Comment (Domejko) [reply settings/prod_settings.py:43]: Yes but while **local** is already defined and set in stone, **prod** still have a possibility of been `None`. Both of them need to set with the same type otherwise `mypy` is throwing an error so we can leave it with `list[str | None]` or set `list[str]` and use `# type: ignore`

Comment (TreyWW) [reply settings/prod_settings.py:43]: Ah I see. I guess for now we can keep it as list[str | None] like you suggested then.

File: backend/views/core/invoices/edit.py
Comment (Domejko) [reply backend/views/core/invoices/edit.py:143]: Ok, so I'll use there `# type: ignore`

File: settings/helpers.py
Comment (Domejko) [reply settings/helpers.py:108]: I think it's due to `SingleTemplatedEmailContent` and that `template_data` has also `dict` type, because when i left it with only `str` type error was gone from `data_str` but it moved to next line 109 and complain that `data.content.get("template_data")` return type is `str | None` instead of `str`.

So to keep type `str` we need to put `# type: ignore[assignment]` on line 108 and 109 because 109 got return type of `dict[Any, Any] | str | None` in this case.

Comment (TreyWW) [reply settings/helpers.py:108]: Ah so I guess instead of data.content.get("template_data") we was meant to use data_str? I just didn't use the variable?

Comment (Domejko) [reply settings/helpers.py:108]: No `data_str` is used once later on in content of `send_email` and we need `data.content.get("template_data")` to assign it's value to `data_str` when it's just a string but `mypy` assume that `data.content.get("template_data")` might be empty and can return only `None` so he complains that `data_str` demands type `str` but `.get` return type is `str | None`.

I would suggest to leave `data_str` with out type since `SingleTemplatedEmailContent` is already covering `template_data` as `str`, we might only add in `SingleTemplatedEmailContent` to template_data `dict[str, str] | str` instead of just `dict | str` so it will cover that the content of dict should also be a string.

Comment (Domejko) [reply settings/helpers.py:108]: Let me know which option would you prefer and I will push the changes.

Comment (TreyWW) [reply settings/helpers.py:108]: > I would suggest to leave data_str with out type since SingleTemplatedEmailContent is already covering template_data as str, we might only add in SingleTemplatedEmailContent to template_data dict[str, str] | str instead of just dict | str so it will cover that the content of dict should also be a string.

Yeah that sounds good to me, apologies for the late reply.

It's odd cause
```py
data_str: str | Any = (
                    data.content.get("template_data")
                    if isinstance(data.content.get("template_data"), str)
                    else json.dumps(data.content.get("template_data"))
)
```
will always be a string no matter what, bit odd that mypy flags that. Thanks for the resolution though

Comment (Domejko) [reply settings/helpers.py:108]: No problem ;) When I'll be on my working station I'll make this change, push it and will let you know when it's ready.

**Comment by Domejko** [comment]
New commit with requested changes have been pushed.

**Review by TreyWW — APPROVED** [review]

**Comment by Domejko** [comment]
So I think I'm going to hold off on correcting errors until #339 will be implemented.

**Comment by TreyWW** [comment]
Thank you

**Comment by Domejko** [comment]
Ok, tomorrow I'll resolve conflicts and as soon as this PR will be merged I will proceed with work on the rest of errors.

**Comment by TreyWW** [comment]
Amazing, sounds good

**Comment by Domejko** [comment]
Regarding changes. If there will come features in the future that will be using `@login_required` decorator then instead of regular `HttpRequest` should be used `AuthenticatedHttpRequest` from *backend/utils/http_utils.py*. This will prevent `mypy` from throwing error about `AnonymousUser` type.

Also what do you think about moving `HtmxHttpRequest` subclass from *backend/types/htmx.py* to *backend/utils/http_utils.py*. Or maybe other way around, moving `AuthenticatedHttpRequest` to *backend/types/* as **authenticated_http.py**.

**Comment by TreyWW** [comment]
All views are login required, we just use middleware instead of @login_required

**Comment by TreyWW** [comment]
As for the second part of the comment, I'm not too sure. I feel like the utils file should be for logic, rather than type definitions. Wherever the HTMX request is we could add a user object to that? And then add a second one if we use htmx without login required? We could talk more on discord if you'd like, may be easier

**Comment by Domejko** [comment]
Yes sure, my username is **domejko_**

**Review by TreyWW — APPROVED** [review]
Amazing PR, thank you!

**Pull request merged**
